src/IO/SnappyReadBuffer.cpp

```diff
@@ -0,0 +1,68 @@
+#include <Common/config.h>
+
+#if USE_SNAPPY
+#include <memory>
+#include <fcntl.h>
+#include <sys/types.h>
+
+#include <snappy.h>
+
+#include <IO/WriteHelpers.h>
+#include <IO/copyData.h>
+#include <IO/WriteBufferFromString.h>
+
+#include "SnappyReadBuffer.h"
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int SNAPPY_UNCOMPRESS_FAILED;
+    extern const int SEEK_POSITION_OUT_OF_BOUND;
+}
+
+SnappyReadBuffer::SnappyReadBuffer(std::unique_ptr<ReadBuffer> in_, size_t buf_size, char * existing_memory, size_t alignment)
+    : BufferWithOwnMemory<SeekableReadBuffer>(buf_size, existing_memory, alignment), in(std::move(in_))
+{
+    WriteBufferFromString wb(compress_buffer);
+    copyData(*in, wb);
+
+    bool success = snappy::Uncompress(compress_buffer.data(), wb.count(), &uncompress_buffer);
```

Comment:
This is not correct to do it this way and also not correct to do it in constructor. Need to override `nextImpl` and implement this logic like in other compression read buffers.

Reply:
ok.

Reply:
I found it difficult to implement stream compress/decompress for snappy.  As far as I know, snappy library (https://github.com/google/snappy) doesn't contain stream compress/decompress API for snappy. Do you have any ideas about this problem? @kssenii